### PR TITLE
Migrating all binaries to the binaries repo

### PR DESCRIPTION
**Pull request by vedpatwardhan**
This PR removes the binaries from the `ivy` repo which would instead be stored on a dedicated repo named `binaries` to enable pulling only relevant binaries to a particular user's interpreter, OS and architecture.

### CODE REVIEW

**Comment by vedpatwardhan** [comment]
I've done the following,

1. Removed all binaries from the `ivy` repo and added them instead to the [binaries](https://github.com/unifyai/binaries) repo with the naming convention same as for pypi wheels for consistency of naming. Also added a `binaries.json` to `ivy` for the skeleton structure. We'd need to add more binaries for different architectures and also different python versions as the current ones are only for python 3.9 and 3.10, but @vaatsalya123 could you please take a look at this part of the changes?
2. Updated the `setup.py` to download the binaries corresponding to a particular architecture with the same preference list of names of binaries as the names of wheels used by pypi (with `sys.tags()` being that preference list) with a spinner, a progress bar and so on. I'm not sure if it's as clean as it can be, @guillesanbri could you please take a look at the changes to `setup.py`?
3. In the rare case of someone wanting to clone repo, checkout to an older release (through the tag) and pulling the binaries corresponding to that release, I've added a "VERSION" environment variable they can set. But for this to be useful, we'd first need to ensure that we also add tags to the `binaries` repo when we make our releases. @RashulChutani could you please take a look at the changes made to `pypi.yml`?

Happy to know about improvements or clarify about any of these changes, thanks!

File: setup.py [hunk 1]
```diff
@@ -18,13 +18,71 @@
 import setuptools
 from setuptools import setup
 from pathlib import Path
+from packaging import tags
```
Comment (guillesanbri) [inline setup.py:21]: had to install this manually

File: setup.py [hunk 2]
```diff
@@ -18,13 +18,71 @@
...
+import os
+import json
+import requests
```
Comment (guillesanbri) [inline setup.py:24]: same with requests

File: setup.py [hunk 3]
```diff
@@ -18,13 +18,71 @@
...
+import itertools
 import re
+from tqdm import tqdm
```
Comment (guillesanbri) [inline setup.py:27]: and same with tqdm

File: setup.py [hunk 4]
```diff
@@ -18,13 +18,71 @@
...
+        r = requests.get(
+            f"https://github.com/unifyai/binaries/raw/{version}/{search_path}",
+            timeout=40,
+        )
+        if r.status_code == 200:
+            if pbar is None:
+                print()
+                print("Downloading binaries ...")
+                pbar = tqdm(total=len(paths))
+            with open(path, "wb") as f:
```
Comment (guillesanbri) [inline setup.py:78]: can't pass over this point, I get:
```
          with open(path, "wb") as f:
      FileNotFoundError: [Errno 2] No such file or directory: 'ivy/compiler/utils/III.so'
```

Seems to work if I create the utils dir manually, so I'm assuming we are missing `os.makedirs(os.path.dirname(path), exist_ok=True)` before `open`

File: setup.py [hunk 5]
```diff
@@ -18,13 +18,71 @@
...
+                f.write(r.content)
+            terminate = path == paths[-1]
+            pbar.update(1)
+        else:
```
Comment (guillesanbri) [inline setup.py:82]: maybe worth adding an error message if the connection hasn't been successful :)

**Review by guillesanbri — COMMENTED** [review]
Left a few comments, I tried installing in a fresh environment, which I assume is the cause of the missing packages, will need to sort this out nonetheless :sweat_smile: 

I think the `makedirs` call should fix the other problem I was facing. I tried on a fresh conda environment, let me know if something else should be specially tested :+1:

File: setup.py (hunk 4 above)
Comment (vedpatwardhan) [reply setup.py:73]: done, thanks for pointing that out!

File: setup.py (hunk 1 above)
Comment (vedpatwardhan) [reply setup.py]: replaced it with a built-in alternative, thanks for pointing that out!

**Comment by vedpatwardhan** [comment]
done with the changes, please take a look whenever you're free, thanks!

File: setup.py [hunk 6]
```diff
@@ -18,13 +18,72 @@
...
+            terminate = path == paths[-1]
+        except request.HTTPError:
+            break
+
+for path in paths:
+    if not os.path.exists(path):
+        print(
+            f"Could not download {path}. Check out the unifyai/binaries README for"
+            " the supported configurations!"
+        )
```
Comment (guillesanbri) [inline setup.py:84]: lgtm!

**Review by guillesanbri — APPROVED** [review]

File: setup.py (hunk 5 above)
Comment (guillesanbri) [reply setup.py]: Ah got it yeah, this works fine then, thanks Ved! :+1:

File: requirements/requirements.txt [hunk 7]
```diff
@@ -7,6 +7,8 @@ packaging
 nvidia-ml-py # mod_name=pynvml
 diskcache
 google-auth # mod_name=google.auth
+urllib3<2.0 # for version conflict with google-auth
 requests
 pyvis
 dill
+tqdm
```
[the remark this answers is not in the document]
Comment (vedpatwardhan) [reply requirements/requirements.txt]: reverted the changes to `requirements` and `optional` files 👍
actually `tqdm` is being used in the `determine_test_coverage.py` file, probably @RashulChutani could comment on it's usability 😅

**Comment by vedpatwardhan** [comment]
> 1. Removed all binaries from the `ivy` repo and added them instead to the [binaries](https://github.com/unifyai/binaries) repo with the naming convention same as for pypi wheels for consistency of naming. Also added a `binaries.json` to `ivy` for the skeleton structure. We'd need to add more binaries for different architectures and also different python versions as the current ones are only for python 3.9 and 3.10, but @vaatsalya123 could you please take a look at this part of the changes?
> 3. In the rare case of someone wanting to clone repo, checkout to an older release (through the tag) and pulling the binaries corresponding to that release, I've added a "VERSION" environment variable they can set. But for this to be useful, we'd first need to ensure that we also add tags to the `binaries` repo when we make our releases. @RashulChutani could you please take a look at the changes made to `pypi.yml`?

Hey @vaatsalya123 and @RashulChutani, could you please take a look? Given it's directly linked to the binaries, I'm refraining from merging without multiple opinions. Thanks!

File: requirements/requirements.txt (hunk 7 above)
[the remark this answers is not in the document]
Comment (guillesanbri) [reply requirements/requirements.txt]: ah missed that, true!

[the remark this answers is not in the document]
Comment (vedpatwardhan) [reply requirements/requirements.txt]: no worries!

**Comment by vedpatwardhan** [comment]
> Overall, it looks good to me. However, the binaries code is better suited to be in the release workflow instead of pypi.yml, as that is dedicated for PyPI deployment.

Thanks for the review, I've basically added a `binaries.yml` for adding the tags to the binaries repo and added it as a job to `release.yml`, happy to know your thoughts 😄

**Comment by vaatsalya123** [comment]
@vedpatwardhan All look's perfect to me , Just so I'm sure did you test the full pipeline in action to confirm it's working or are we yet to do this?

**Comment by vedpatwardhan** [comment]
I'll get this merged for now, let me know if there's anything to be updated with the `binaries.yml` workflow @RashulChutani I'll get that fixed. Ofc given that the `binaries.yml` workflow wouldn't be used until we make a release, it would lead to any immediate problems so merging the PR for now. Thanks all 😄

**Pull request merged**